tools/data_gen/src/main.rs

```diff
@@ -0,0 +1,173 @@
+            }
+            SizeVec::Quad(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 0, "data", &v)
+            }
+        }
+    }
+}
+
+#[derive(FromArgs)]
+/// Used to read a file
+struct FilePath {
+    /// file path to read data from
+    #[argh(positional, from_str_fn(read_path))]
+    file: Option<PathBuf>,
+}
+
+fn read_path(path: &str) -> Result<PathBuf, String> {
+    Ok(Path::new(path).into())
+}
+
+fn main() -> CalyxResult<()> {
+    let fp: FilePath = argh::from_env();
+
+    let lib_path: &Path = &Path::new("../../../calyx");
+
+    let ws = frontend::Workspace::construct(&fp.file, lib_path)?;
+    let ctx: ir::Context = (ir::from_ast::ast_to_ir(ws))?;
+
+    let comp = ctx
+        .components
+        .into_iter()
+        .find(|comp| comp.name == ctx.entrypoint)
+        .expect("No top-level component found.");
+
+    let data_vec: Vec<CellData> = comp
+        .cells
+        .iter()
+        .filter_map(|cell| get_data(cell))
+        .collect();
+
+    let mut map = Map::new();
+
+    for CellData {
+        name,
+        width,
+        mut sizes,
+    } in data_vec
+    {
+        let json_comp: serde_json::Value = gen_comp(&mut sizes, width);
+        map.insert(name, json_comp);
+    }
+
+    let json_map: Value = map.into();
+    println!("{}", json_map.to_string());
+    Ok(())
+}
+
+//generates a json value associated with sizes_vec and width
+fn gen_comp(sizes_vec: &mut Vec<u64>, width: u64) -> serde_json::Value {
+    let initial: SizeVec =
+        SizeVec::Single(vec![0; sizes_vec.pop().unwrap().try_into().unwrap()]);
+    sizes_vec.reverse();
+    let data_vec: SizeVec =
+        sizes_vec.iter().fold(initial, |acc, x| accumulate(*x, acc));
+    let data = serde_json::to_value(data_vec).unwrap();
+    json!({
+        "data": data.get("data").unwrap(),
+        "format": {
+            "numeric_type": "bitnum",
+            "is_signed": false,
+            "width": width,
+        }
+    })
+}
+
+//creates a vector consisting of i v's.
+fn add_dimension<T: std::clone::Clone>(i: u64, v: Vec<T>) -> Vec<Vec<T>> {
+    vec![v; i.try_into().unwrap()]
+}
+
+//function to help build vectors of multiple dimensions using fold()
+fn accumulate(i: u64, v: SizeVec) -> SizeVec {
+    match v {
+        SizeVec::Single(v) => SizeVec::Double(add_dimension(i, v)),
+        SizeVec::Double(v) => SizeVec::Triple(add_dimension(i, v)),
+        SizeVec::Triple(v) => SizeVec::Quad(add_dimension(i, v)),
+        SizeVec::Quad(_) => panic!("trying to add dimension to 4d"),
+    }
+}
+
+//Takes in cell, returns the reference to the cell
+fn get_ref(cell: &Rc<RefCell<ir::Cell>>) -> Ref<ir::Cell> {
+    let ref_cell: &RefCell<ir::Cell> = cell.borrow();
+    let final_cell: Ref<ir::Cell> = ref_cell.borrow();
+    final_cell
+}
+
+//Returns Some(CellData)) if cell is a std_mem cell, None otherwise
+fn get_data(cell: &Rc<RefCell<ir::Cell>>) -> Option<CellData> {
+    let final_cell = get_ref(cell);
```

Comment:
You don't need this function. I think you can just do `&*cell.borrow()`.

Reply:
For some reason this fix wasn't working for me. I think it's because I have to call borrow twice? Once to get the RefCell and then once to get the Ref?

Reply:
See my comment on the previous helper function. Once resolved, this can also be removed